---
title: "Building an epigenetic clock based on methods and data from Hannum et al"
output: html_document
date: "`r Sys.Date()`"
Author: "Solomon Lalala"
---

# Description

###### Aging is complex biological process with decline in physiology and increased vulnerability to diseases. Developing intervention on aging and aging related diseases involves measuring biological age and aging rate at the molecular level. Aging clocks are machine learning models trained with molecular features such as genes or proteins to estimate an individual's age. Epigenetic aging clock build upon DNA methylation pattern that changes with aging. Epigenetic aging clock is robust in predicting age across various tissue types and can measure epigenetic changes shared between aging and many cancers. This project uses existing methods developed by Hannum et al. to train an epigenetic aging clock and aims to predict the age and aging rate of individuals using public available DNA methylation data.

# Summary

##### An epigenetic clock was trained using elastic net regression with lambda = 0.4868, alpha = 0.5, and 48 CpG markers using 493 samples of DNA methylation data from GSE40279. Evaluated with 163 samples, the model had a correlation coefficient (r) of 0.89 between actual age and predicted age and root mean squared error (rmse) of 6.77 years. Aging rate was calculated by dividing predicated age by actual age. The accuracy of the model is high but is not as good as the Hannum's clock. It would be good to train the model with the whole dataset of 656 samples and benchmark using other datasets.

# Setup

1.  Install the following packages "tidyverse", "knitr", "caret", "glmnet", "ggplot2", "GEOquery", "foreach", "doParallel".

```{r Install packages, eval = FALSE}
install.packages(c("tidyverse", "knitr", "caret", "glmnet", "ggplot2", "GEOquery", "foreach", "doParallel"))
```

2.  Run the following codes

## Import scripts

```{r setup, echo=TRUE, message=FALSE, warning=FALSE}
# Helper functions
source("scripts/helper.R")
```

## Set directory and dataset id

```{r set directory and dataset id}
data_dir <- "data" 
fig_dir <- "figures" 
gse_accession <- "GSE40279" 
```

# Collect data

**Primary dataset for training and testing the model**:[**GSE40279**](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE40279)

```{r parse data, eval = FALSE}
# Parse relevant metadata and DNA methylation matrix from GEO database. The metadata and DNA methylation matrix are saved in data_dir, and a list containing the two dataframe is returned.
data_list <- parse_geo(dest_dir = data_dir, gse_id = gse_accession, transpose = TRUE, save = TRUE)
```

# Process data

## Extract metadata and merge with DNA methylation data

-   Relevant metadata is extracted from the metadata dataframe and merged with the transposed DNA methylation matrix.

```{r merge data}
merged <- extract_merge(gse_id = gse_accession, dest_dir = data_dir, save = TRUE)
```

## Viewing the data

-   sample_id: GSM id for each sample. Each sample is collected from the whole blood of an individual.
-   age (target): The age of the individual. It is the prediction target.
-   CpG marker (predictor): Those columns starting with cg are CpG marker. The values are continuous variables between 0 and 1, representing the frequency of the methylation of the CpG marker across the population of the blood cells of the given individual.
-   Ethnicity and sex were merged but excluded from model training.

```{r view data}
dim(merged)
# select 6 samples and 10 columns
view_merged(file.path(data_dir, paste0(gse_accession, "merged.rds")))
```

### View age distribution of merged data

The majority of the samples are between 50 and 80 years old.

```{r age distribution}
p_age_merged <- plot_age_distribution(
  data = merged,
  dest_dir = fig_dir, 
  save = TRUE)
p_age_merged
```

### View sex distribution of merged data

-   Sex is not exactly equal in the merged data with 20 more female than male.

```{r sex distribution}
table(merged$sex)
```

## Split merged data into train and test sets

-   Merged data is split into train (0.75) and test (0.25) sets using stratified sampling by age and gender.

```{r split train and test}
train_test_list <- split_train_test(
  dataset = merged, 
  train_frac = 0.75, 
  stratifier = "age", 
  dest_dir = data_dir, 
  save = TRUE)
train_set <- train_test_list[[1]]
test_set <- train_test_list[[2]]
test_meta <- test_set %>%
  select(sample_id, sex)
saveRDS(test_meta, file = file.path(data_dir, "test_meta.rds"))
```

### Check subset data distribution

```{r subset distribution}
# Check age distribution
p_age_train <- plot_age_distribution(data = train_set, dest_dir = fig_dir, save = TRUE)
p_age_train
p_age_test <- plot_age_distribution(data = test_set, dest_dir = fig_dir, save = TRUE)
p_age_test
table(train_set$sex)
table(test_set$sex)
```

## split data into predictor and target

-   The predictor set is the DNA methylation matrix, and the target set is the age of the individual.

```{r split predictor and target, eval = FALSE}
X_train <- train_set %>%
  select(-sample_id, -age, -ethnicity, -sex) 
y_train <- train_set$age

X_test <- test_set %>%
  select(-sample_id, -age, -ethnicity, -sex)
y_test <- test_set$age
```

```{r save splited data, eval = FALSE}
saveRDS(X_train, file = file.path(data_dir, "X_train.rds"))
saveRDS(y_train, file = file.path(data_dir, "y_train.rds"))
saveRDS(X_test, file = file.path(data_dir, "X_test.rds"))
saveRDS(y_test, file = file.path(data_dir, "y_test.rds"))
```

## standardize data

```{r center and scale data, eval = FALSE}
# obtain mean and sd
preProcValues <- preProcess(as.matrix(X_train), method = c("center", "scale"))

# Apply the transformation to the data
X_train_standardized <- predict(preProcValues, as.matrix(X_train))
X_test_standardized <- predict(preProcValues, as.matrix(X_test))
```

```{r save standardized data, eval = FALSE}
# save prepro_values and standardized data
saveRDS(preProcValues, file = file.path(data_dir, "preProcValues.rds"))
saveRDS(X_train_standardized, file = file.path(data_dir, "X_train_standardized.rds"))
saveRDS(X_test_standardized, file = file.path(data_dir, "X_test_standardized.rds"))
```

# Train model

-   The model is trained using Elastic Net regression, a combination lasso regression and ridge regression. Lasso regression adds a penalty lambda proportional to the absolute value of the coefficients beta, which shrinks some coefficients to zero to eliminate less important features, while ridge regression adds a penalty lambda proportional to the square of the coefficients beta, which shrinks coefficients towards zero and make the features contribute less to prediction.

-   In the DNA methylation profiles, the number of CpG markers (predictors) is much more than the number of samples, and elastic net regression can reduce the number of CpG markers and select those markers contributing more to age prediction.

-   493 samples were used to train the model. 

-   Elastic net regression introduces a hyperparameter alpha to control the weight of lasso regression (alpha = 1) and ridge regression (alpha = 0).

-   n_cores is the number of cores to use for parallel processing.

-   n_fold is the number of folds for cross-validation.

-   iterations is the number of iterations for the bootstrap sampling.
## Set parameters
```{r set parameters}
alpha = 0.5
n_cores <- parallel::detectCores() - 1
n_fold = 10
iterations = 500
model_params <- list(
  n_cores = n_cores,
  alpha = alpha,
  n_fold = n_fold,
  iterations = iterations
)
```

```{r save parameters, eval = FALSE, echo = FALSE}
saveRDS(model_params, file = file.path(data_dir, "model_params.rds"))
```

## Find optimal parameters

### Cross-validation

-   Cross-validation is used to find the optimal parameter lambda, which controls the level of penalty applied to the coefficients.

```{r cv in r, eval = FALSE}
# Run locally
Rscript -e "source('scripts/cv.R')"
```

```{bash cv job, eval = FALSE}
# OR submit a job to SCC. You need to have access to SCC and set up the bash script
qsub scripts/cv.sh
```

### Find lambda

-   y axis is the mean squared error (MSE) of the model, and x axis is the log of lambda. The top row shows the number of non-zero coefficients. The dash vertical line on the left shows the optimal lambda (lambda.min), and the dash vertical line on the right shows the lambda that gives one standard error (lambda.1se).

```{r find lambda}
cv_model <- readRDS(file.path(data_dir, "cv_model.rds"))
plot(cv_model)
```

```{r lambda min}
lambda_min <- cv_model$lambda.min
lambda_min
```

```{r save cv model, eval = FALSE}
ggsave(
  filename = file.path(fig_dir, "cv_model_lambda.png"),
  plot = plot(cv_model),
)
saveRDS(lambda_min, file = file.path(data_dir, "lambda_min.rds"))
```

## Select feature

### Bootstrap

-   For feature selection, the train dataset was resampled with replacement 500 times, and a model was built for each bootstrap.

```{bash Bootstrap job, eval = FALSE}
# Submit a job to SCC. You need to have access to SCC and set up the bash script
qsub scripts/bootstrap.sh
```

```{r Bootstrap in r, eval = FALSE}
# OR run locally. would be slow or could crash
Rscript -e "source('scripts/bootstrap.R')"
```

### Select feature

-   CpG markers that were present in more than half of all bootstraps were selected to train the final model.

```{r Select feature}
bootstrap_res <- readRDS(file.path(data_dir, "bootstrap_results.rds"))
selected_feature <- select_feature(
  bootstrap_results = bootstrap_res, 
  threshold = 0.5,
  dest_dir = data_dir,
  save = TRUE)
```

#### View selected feature

-   48 CpG markers were selected.

```{r view selected feature}
selected_feature
```

## Train final model

### Load data

```{r load data for final model}
# Load data
X_train_standardized <- readRDS(file.path(data_dir, "X_train_standardized.rds"))
y_train <- readRDS(file.path(data_dir, "y_train.rds"))
X_test_standardized <- readRDS(file.path(data_dir, "X_test_standardized.rds"))
y_test <- readRDS(file.path(data_dir, "y_test.rds"))
```

### Select features in training data

```{r}
# Select features in training data and testing data
X_train_standardized_final <- X_train_standardized[, selected_feature]
X_test_standardized_final <- X_test_standardized[, selected_feature]
```

### Fit model

-   alpha = 0.5, lambda = 0.4868

```{r train final model}
set.seed(123)
final_model <- glmnet(
    X_train_standardized_final, 
    y_train, 
    alpha = alpha, 
    lambda = lambda_min, 
    type.measure = "mse", 
    family = "gaussian", 
    standardize = FALSE) 
```

```{r save final model, eval = FALSE}
saveRDS(final_model, file = file.path(data_dir, "final_model.rds"))
```

# Evaluate model

-   163 samples were used to evaluate the model.
-   The correlation coeffcient (r) and the root mean squared error (rmse) were calculated.

```{r evaluate model}
evaluation <- evaluate_model(
  model = final_model, 
  X_test = X_test_standardized_final, 
  y_test = y_test, 
  dest_dir = data_dir)
```

## Visualize evaluation results

-   The predicted age is plotted against the actual age. The red line is the line of best fit (predicted age = actual age).
-   r is the correlation coefficient between actual age and predicted age.

```{r view evaluation}
eval_res <- evaluation$results_df
eval_metrics <- evaluation$metrics

plot_eval <- plot_evaluation(
  results = eval_res, 
  metrics = eval_metrics, 
  dest_dir = fig_dir)
plot_eval
```

## Calculate aging rate

##### Aging rate was calculated by dividing predicated age by actual age.

```{r}
eval_res <- eval_res %>%
  mutate(Aging_Rate = Predicted_Age / Actual_Age) %>%
  left_join(test_meta, by = c("Sample" = "sample_id")) %>%
  mutate(sex = recode(sex, "F" = "Female", "M" = "Male"))

amar <- ggplot(eval_res, aes(x = Aging_Rate, color = sex)) +
  geom_density(size = 1.2) +
  scale_color_manual(values = c("Female" = "black", "Male" = "red")) +
  labs(x = "Aging Rate (AMAR)",
       y = "Density") +
  theme_classic(base_size = 14) +
  theme(legend.title = element_blank())
print(amar)
ggsave(
  filename = file.path(fig_dir, "aging_rate.png"),
  plot = amar
)
```

# Reference

###### Anastasiadi D, Piferrer F. 2023. Bioinformatic analysis for age prediction using epigenetic clocks: Application to fisheries management and conservation biology. Front Mar Sci. 10. <doi:10.3389/fmars.2023.1096909>.

###### Farrell CP. A Simple Epigenetic Clock Using Python and SciKit-Learn. Colin P Farrell. <https://colinpfarrell.com/tag-ec-tutorial/>.

###### Hannum G, Guinney J, Zhao L, Zhang L, Hughes G, Sadda S, Klotzle B, Bibikova M, Fan J-B, Gao Y, et al. 2013. Genome-wide Methylation Profiles Reveal Quantitative Views of Human Aging Rates. Mol Cell. 49(2):359â€“367. <doi:10.1016/j.molcel.2012.10.016>.

###### OpenAI. (2025). ChatGPT (Mar 2025 version). Accessed via <https://chat.openai.com>

```{r session-info}
sessionInfo()
```
